src/dbnode/storage/index.go

```diff
@@ -1436,3 +1440,230 @@ func (shards dbShards) IDs() []uint32 {
 	}
 	return ids
 }
+
+func (i *nsIndex) AggregateQuery(
+	ctx context.Context,
+	query index.Query,
+	opts index.AggregateQueryOptions,
+) (index.AggregateResults, error) {
+	// Capture start before needing to acquire lock.
+	start := i.nowFn()
+
+	i.state.RLock()
+	if !i.isOpenWithRLock() {
+		i.state.RUnlock()
+		return nil, errDbIndexUnableToQueryClosed
+	}
+
+	// Track this as an inflight query that needs to finish
+	// when the index is closed.
+	i.queriesWg.Add(1)
+	defer i.queriesWg.Done()
+
+	// Enact overrides for query options
+	opts.QueryOptions = i.overriddenOptsForQueryWithRLock(opts.QueryOptions)
+	timeout := i.timeoutForQueryWithRLock(ctx)
+
+	// Retrieve blocks to query, then we can release lock
+	// NB(r): Important not to block ticking, and other tasks by
+	// holding the RLock during a query.
+	blocks, err := i.blocksForQueryWithRLock(xtime.NewRanges(xtime.Range{
+		Start: opts.StartInclusive,
+		End:   opts.EndExclusive,
+	}))
+
+	// Can now release the lock and execute the query without holding the lock.
+	i.state.RUnlock()
+
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		deadline = start.Add(timeout)
+		wg       sync.WaitGroup
+
+		// Results contains all concurrent mutable state below.
+		results = struct {
```

Comment:
can you extract this into a separate struct and reuse it across both places

Reply:
Sure, had something like that originally while trying to jam it into a combined function :)